### DISCUSSION TITLE
Using Target Rate Sensors with Solis Cloud Integration

### DISCUSSION THREAD

**Discussion by CraigCallender** [BottlecapDave/HomeAssistant-OctopusEnergy#809] [Show and tell]
So I borrored a good bit from #284, but the Solis Inverter has three timeslots that can be configured to force charging from the grid.  I followed the tutorial here to setup the Target Rate setting in the Octopus integration.  

Originally, I configured the automation to look at the binary_sensor for when my `binary_sensor.octopus_energy_target_battery_charge_slots` changed state to `on`, I would just start charging and the start and end times were set to the `sensor.octopus_energy_electricity_18l2077993_1900013281497_current_rate` `start` and `end` attributes. I didn't have to worry about the off state since I'm setting the end time.  And I was assuming the `on` state would come on at the same time as the `current_rate.start`.  

This worked well, but I had one more issue... I don't run HomeAssistant 24/7.  I will, but I don't now.  Also, Solis Cloud will often go down, so I wanted the ability to set all three times as soon as they're calculated in case something happens where I can't set it in real time.

In the end, I ended up with this.  You'll probably need to use the YAML editor as I had issues with the `true/false` states being interpretted as strings when using the UI mode to setup the trigger and condition.  You can always use the UI mode then just get rid of the `""` around `true` and `false` if that's easier.

Trigger - Look for the `target_times_last_evaluated` to change (this seems to only change when new prices come out).
```
trigger:
  - platform: state
    entity_id:
      - binary_sensor.octopus_energy_target_battery_charge_slots
    attribute: rates_incomplete
    from: true
    to: false
  - platform: state
    entity_id: binary_sensor.octopus_energy_target_battery_charge_slots
    to: "on"
```
The two triggers are important to me because I also hijack the first charge time slot if electricity rates every go under 0.02 GBP / kWh.  So in case that happened, I just reset the values whenever the state of the binary_sensor turns on.

I've added a condition as we can't run this if `rates_incomplete` is set to `true`:
```
condition:
  - condition: state
    entity_id: binary_sensor.octopus_energy_target_battery_charge_slots
    attribute: rates_incomplete
    state: false
```

Action - Use the [SolisCloud integration](https://github.com/hultenvp/solis-sensor) to set the inverter's three charge slots.  You can only do this with a service, so it's a bit verbose.  Also, the mixed strings in the `config` for the service are like that for some reason.
```
action:
  - service: pyscript.solis_control
    data:
      days:
        - chargeCurrent: "50"
          dischargeCurrent: "0"
          chargeStartTime: >-
            {% set dt_string =
            as_timestamp(state_attr('binary_sensor.octopus_energy_target_battery_charge_slots',
            'target_times')[0].start) | timestamp_local | as_datetime %} {{
            dt_string.strftime("%H") }}:{{ dt_string.strftime("%M") }}
          chargeEndTime: >-
            {% set dt_string =
            as_timestamp(state_attr('binary_sensor.octopus_energy_target_battery_charge_slots',
            'target_times')[0].end) | timestamp_local | as_datetime %} {{
            dt_string.strftime("%H") }}:{{ dt_string.strftime("%M") }}
          dischargeStartTime: "00:00"
          dischargeEndTime: "00:00"
        - chargeCurrent: "50"
          dischargeCurrent: "0"
          chargeStartTime: >-
            {% set dt_string =
            as_timestamp(state_attr('binary_sensor.octopus_energy_target_battery_charge_slots',
            'target_times')[1].start) | timestamp_local | as_datetime %} {{
            dt_string.strftime("%H") }}:{{ dt_string.strftime("%M") }}
          chargeEndTime: >-
            {% set dt_string =
            as_timestamp(state_attr('binary_sensor.octopus_energy_target_battery_charge_slots',
            'target_times')[1].end) | timestamp_local | as_datetime %} {{
            dt_string.strftime("%H") }}:{{ dt_string.strftime("%M") }}
          dischargeStartTime: "00:00"
          dischargeEndTime: "00:00"
        - chargeCurrent: "50"
          dischargeCurrent: "0"
          chargeStartTime: >-
            {% set dt_string =
            as_timestamp(state_attr('binary_sensor.octopus_energy_target_battery_charge_slots',
            'target_times')[2].start) | timestamp_local | as_datetime %} {{
            dt_string.strftime("%H") }}:{{ dt_string.strftime("%M") }}
          chargeEndTime: >-
            {% set dt_string =
            as_timestamp(state_attr('binary_sensor.octopus_energy_target_battery_charge_slots',
            'target_times')[2].end) | timestamp_local | as_datetime %} {{
            dt_string.strftime("%H") }}:{{ dt_string.strftime("%M") }}
          dischargeStartTime: "00:00"
          dischargeEndTime: "00:00"
      config:
        secret: <<<API Key>>
        key_id: "<<Application ID>>"
        username: << username to log into solis cloud >>
        password: << password to log into solis cloud >>
        plantId: "<< Plant ID >>"
```

Hopefully that helps the next person out trying to get this configured with the Octopus integration.

**Comment by CraigCallender** [comment]
Updated on 2 April 2024 @ 16:48 BST (+01:00) - I added converting the charge slot times from GMT to local time as the integration gives them in GMT but the inverter needed them in local time.

**Comment by CraigCallender** [comment]
Updated at 4 April 2024 @ 08:43 BST (+01:00) to fix the trigger and conditions.  I found that the old ones were too noisy and resulted in multiple sets.

**Comment by usbrit** [comment]
@CorithMalin82 Can you confirm that you are able to change the inverter settings with the script? I've seen lots of posts about "pulling " data from the inverter, but not controlling (without Modbus).

I can't see how:
chargeCurrent: "50"
          dischargeCurrent: "0"
          chargeStartTime: >-
            {% set dt_string =
            as_timestamp(state_attr('binary_sensor.octopus_energy_target_battery_charge_slots',
            'target_times')[0].start) | timestamp_local | as_datetime %} {{
            dt_string.strftime("%H") }}:{{ dt_string.strftime("%M") }}
          chargeEndTime: >-
            {% set dt_string =
            as_timestamp(state_attr('binary_sensor.octopus_energy_target_battery_charge_slots',
            'target_times')[0].end) | timestamp_local | as_datetime %} {{
            dt_string.strftime("%H") }}:{{ dt_string.strftime("%M") }}
          dischargeStartTime: "00:00"
          dischargeEndTime: "00:00"
sets any values on the API?

Thanks!

**Reply by CraigCallender** [reply]
Hi @usbrit. I can confirm that this has been working. Just to confirm though, I’m able to control the inverter with this as the script is provided by the [solis-sensor](https://github.com/hultenvp/solis-sensor) HACS integration. Without that setup, this script doesn’t exist. 

Confirmed by logging into SolisCloud and looking at the inverter settings. Also, I can see on SolisCloud that the charge times each night are different.

**Reply by jayeshgarg** [reply]
@usbrit I think this is the script that you need to setup to make the remote config calls
https://github.com/stevegal/solis_control
Also, you need to submit a ticket to enable remote control. I raised the below queries to them.


I have solis S5-EH1p5K-L inverter. It is registered on my email id mentioned in this ticket. I want to know a few things.
1) How can I do remote configurations? currently, it's only possible to do this on the inverter itself
2) Why does time run 1 hour before the actual time. On the inverter panel the time it shows is 1 hour earlier than the actual time. I tried fixing it, but as soon as I save and exit the settings, it goes back to the original value.
3) where on the inverter/solid cloud can I see the max capacity of my batteries. Currently I have attached 2 Weco 5kW batteries, and it only shows the % value on solis cloud. I want to see 10kW or 5kW*2
Please help me understand the product better.
Thank you

RESPONSE
1) We have enabled remote control option for your account. Kindly logout and login
2) We have made DST ON ,so it will fetch automatically when the time get change 
3) You cannot  see the Battery capacity in solis cloud  or on inverter or mobile app


NOTE: I have just found this script and thread, so I have not yet tried it myself. I did try the SolisCloud based remote control and that worked.

**Reply by CraigCallender** [reply]
@jayeshgarg - Thank you.  It seems I was mixing up `solis-sensor` with `solis-control`.  :)

**Reply by usbrit** [reply]
That makes more sense! Thanks to all, I'll give it a go.

**Reply by usbrit** [reply]
@CorithMalin82 - So I've got the pyscript installed but am not sure how to call it using the config? Could you explain that? Sorry, I'm pretty new to python on HA. Thanks

**Reply by usbrit** [reply]
Now up and running, thanks!

**Reply by usbrit** [reply]
Nest problem, I can send data to the inverter (Yay!)
Now can I ask where "binary_sensor.octopus_energy_target_battery_charge_slots" comes from?
Is this part of the original OctopusEnergy integration?
Thanks.

**Reply by jayeshgarg** [reply]
@usbrit first of all congrats on your success.

Yes octopus is an energy provider so this belongs to their HA integration to fetch the tiered rate so that schedule can be set dynamically.

Sorry I was not home so couldn't answer to your earlier posts.

**Comment by usbrit** [comment]
OK, already have this setup, so just need to find the correct attributes to
link to the script.

**Reply by usbrit** [reply]
Am I right in thinking you can only set the first 1-1/2 hours of slots with this code? (30min+30min+30min)?

I tried to use a continuous 3 hr target to get a start and end with the best 3 hour in between but I still get 6 consecutive start and end times, so can only use the first 1-1/2 hours anyway (based on 3 `on` and 3 `off` times available in the Solis setup)

I'd really like to charge for the best continuous 3 hour period

```
Name
best_3c_charge_slot
Hours
3
Type
Continuous
Mpan
XXXXXXXXXXX
Start time
18:00
End time
12:00
Rolling target
false
Last rates
false
Target invert target rates
false
Kind
target_rate
Account ID
XXXXXXXX
Is target export
false
Rates incomplete
false
Target times
- value_inc_vat: 0.147735
start: '2024-04-18T00:00:00+00:00'
end: '2024-04-18T00:30:00+00:00'
tariff_code: E-1R-AGILE-23-12-06-A
is_capped: false
- value_inc_vat: 0.160755
start: '2024-04-18T00:30:00+00:00'
end: '2024-04-18T01:00:00+00:00'
tariff_code: E-1R-AGILE-23-12-06-A
is_capped: false
- value_inc_vat: 0.165375
start: '2024-04-18T01:00:00+00:00'
end: '2024-04-18T01:30:00+00:00'
tariff_code: E-1R-AGILE-23-12-06-A
is_capped: false
- value_inc_vat: 0.15939
start: '2024-04-18T01:30:00+00:00'
end: '2024-04-18T02:00:00+00:00'
tariff_code: E-1R-AGILE-23-12-06-A
is_capped: false
- value_inc_vat: 0.15435
start: '2024-04-18T02:00:00+00:00'
end: '2024-04-18T02:30:00+00:00'
tariff_code: E-1R-AGILE-23-12-06-A
is_capped: false
- value_inc_vat: 0.14595
start: '2024-04-18T02:30:00+00:00'
end: '2024-04-18T03:00:00+00:00'
tariff_code: E-1R-AGILE-23-12-06-A
is_capped: false
Target times last evaluated
18 April 2024 at 07:19:12
Overall average cost
0.16
Overall min cost
0.15
Overall max cost
0.17
Current duration in hours
0
Next duration in hours
0
Last evaluated
18 April 2024 at 07:29:12
```

**Reply by usbrit** [reply]
Further to this, I managed to use the [0] `start` and [6] `end` to get what I wanted. However I now cannot format the time correct to "01:30" but am getting:
```
chargeStartTime: >-
         01:00
```
See Below:
![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/53425475/09b26147-a98f-4678-b48b-73cc8a28d5a7)

Any help with the format appreciated.
Thanks

**Reply by CraigCallender** [reply]
Are you sure you should be getting `01:30`?  From the previous message, I see a start time of `00:00` in GMT, which should result in `01:00` in BST.  Maybe I'm missing something and there's a reason you'd want to start your charge at the end of the first session?

Also, you're correct that this doesn't work for more than 3 non-contiguous cheap sessions.  It probably wouldn't be too hard to expand this with a loop and to loop through the sessions, discarding any that have passed.  Alternatively, you could _just_ rely on the `octopus_energy_target_battery_charge_slots` event, which fires every time a charge session starts.  Then you can set the start time to now() and the end time to now+30 min.  

I chose to not rely on this incase SolisCloud goes down I wanted to have the best chances of not missing the charge time.

**Comment by usbrit** [comment]
> Are you sure you should be getting `01:30`? From the previous message, I see a start time of `00:00` in GMT, which should result in `01:00` in BST. Maybe I'm missing something and there's a reason you'd want to start your charge at the end of the first session?
> 

I changed the target sensor to 3 hours continuous since the original message, so times are different.

> Also, you're correct that this doesn't work for more than 3 non-contiguous cheap sessions. It probably wouldn't be too hard to expand this with a loop and to loop through the sessions, discarding any that have passed. Alternatively, you could _just_ rely on the `octopus_energy_target_battery_charge_slots` event, which fires every time a charge session starts. Then you can set the start time to now() and the end time to now+30 min.
> 
> I chose to not rely on this incase SolisCloud goes down I wanted to have the best chances of not missing the charge time.
Agreed, but like you, don't want to rely on all that data transfer.

What about the format issue? the API doesn't seem to like the
chargeStartTime: >-
         01:00

**Reply by CraigCallender** [reply]
I'm not sure about the format.  I have essentially what you do and the result works.  Some differences I've noted:
* My colouring is very different, but perhaps that's just a configuration issue of the editor?  Or a different version of HA?

Trace result:
![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/125502786/abc1a56a-5b33-4d22-af97-f08b35dcf711)

Step Config:
![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/125502786/ad1ec14c-9a7b-4f88-b87d-d9fa82c5303d)

Maybe it's a regional setting in HA that is treating it differently?  But I agree, that's very strange.
